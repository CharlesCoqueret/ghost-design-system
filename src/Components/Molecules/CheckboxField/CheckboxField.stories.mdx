import { useState } from 'react';
import { Canvas, Meta, ArgsTable, Story } from '@storybook/addon-docs';

import CheckboxField, { ICheckboxFieldProps } from './CheckboxField';

<Meta
  title='Molecule/CheckboxField'
  component={CheckboxField}
  parameters={{
    viewMode: 'docs',
    layout: 'fullscreen',
  }}
  argTypes={{
    dataTestId: {
      table: {
        disable: true,
      },
    },
  }}
/>

# CheckboxField

export const options = [
  { label: 'checkbox label 0', value: 'KEY_0' },
  { label: 'checkbox label 1', value: 'KEY_1', checked: true },
  {
    label: 'checkbox label 2 - which can sometimes be ' + 'very '.repeat(10) + ' long',
    value: 'KEY_2',
    checked: false,
  },
  {
    label: 'checkbox label 3 - which can sometimes be ' + 'extremely '.repeat(50) + ' long',
    value: 'KEY_3',
    checked: undefined,
  },
  { label: 'checkbox label 4 highlighted', value: 'KEY_4', highlighted: true },
  { label: 'checkbox label 5 highlighted and checked', value: 'KEY_5', highlighted: true, checked: true },
];

export const Template = (args) => {
  const [val, setVal] = useState(args.inputValue);
  return <CheckboxField {...args} inputValue={val} onChange={setVal} />;
};

<Canvas withSource='open'>
  <Story name='CheckboxField' args={{ label: 'Checkbox field', inputValue: options }}>
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable of={CheckboxField} sort='requiredFirst' exclude={['dataTestId']} />

### Read Only

<Canvas>
  <Template
    fieldSize={6}
    label='Checkbox field in read only with label size = 4 and field size = 6'
    labelSize={4}
    readOnly={true}
    inputValue={options}
  />
</Canvas>

### Error

<Canvas>
  <Template label='Checkbox field in error' errorMessage='This text is on error' inputValue={options} />
</Canvas>

### Helper

<Canvas>
  <Template
    helperText='Helper text'
    label='Checkbox field with helper and counter'
    mandatory={true}
    inputValue={options}
  />
</Canvas>

### Highlighted

<Canvas>
  <Template
    helperText='Helper text'
    highlighted={true}
    readOnly={true}
    label='Checkbox field highlighted'
    inputValue={options}
  />
</Canvas>

### Disabled

<Canvas>
  <Template disabled={true} helperText='Helper text' label='Checkbox field disabled' inputValue={options} />
</Canvas>

### Inline

<Canvas>
  <Template label='Checkbox field in line' inline={true} inputValue={[options[0], options[1]]} />
</Canvas>
