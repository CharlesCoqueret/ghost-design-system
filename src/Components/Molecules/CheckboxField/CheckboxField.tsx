import React, { ReactElement, Ref } from 'react';
import classnames from 'classnames';

import { GenericField } from '../../Atoms/GenericField';
import { IToggleEntry, CheckboxInput } from '../../Atoms/CheckBoxInput';

export interface ICheckboxFieldProps {
  /** React Container ref (optional, default: undefined) */
  containerRef?: Ref<HTMLDivElement>;
  /** For test purpose only */
  dataTestId?: string;
  /** Disabled field (optional, default: false) */
  disabled?: boolean;
  /** Error message (optional, default: undefined) */
  errorMessage?: string;
  /** Class for the field surrounding the input (optional, default: undefined) */
  fieldClassName?: string;
  /** Size of the field in a 12 column grid (optional, default: undefined) */
  fieldSize?: number;
  /** Helper text (optional, default: undefined) */
  helperText?: string;
  /** Highlighted field and highlight the toggle entries marked as highlight (optional, default: false) */
  highlighted?: boolean;
  /** Inline field (optional, default: false) */
  inline?: boolean;
  /** Class for the input (optional, default: undefined) */
  inputClassName?: string;
  /** Input value */
  inputValue: Array<IToggleEntry>;
  /** Label (optional, default: undefined) */
  label?: string;
  /** Size of the field in a 12 column grid (optional, default: undefined) */
  labelSize?: number;
  /** Mandatory field (optional, default: false) */
  mandatory?: boolean;
  /** Handler of value changes (optional, default: undefined) */
  onChange?: (values: Array<IToggleEntry>) => void;
  /** Read only field (optional, default: false) */
  readOnly?: boolean;
}

/**
 * Checkbox field component
 *
 * Checbox input wrapped in a generic field ( @see GenericField ).
 *
 * Calls @param onChange for every input change.
 *
 */
export const CheckboxField = (props: ICheckboxFieldProps): ReactElement => {
  const {
    containerRef,
    dataTestId,
    disabled,
    errorMessage,
    fieldClassName,
    fieldSize,
    helperText,
    highlighted,
    inline,
    inputClassName,
    inputValue,
    label,
    labelSize,
    mandatory,
    onChange,
    readOnly,
  } = props;

  return (
    <GenericField
      containerRef={containerRef}
      errorMessage={errorMessage}
      fieldClassName={fieldClassName}
      helperText={helperText}
      highlighted={highlighted}
      inline={inline}
      invertInputDescription
      label={label}
      labelSize={labelSize}
      mandatory={mandatory}
      readOnly={readOnly}>
      <CheckboxInput
        className={classnames(
          'field',
          'input-checkbox-field',
          fieldSize && `field-input-size-${fieldSize}`,
          inputClassName,
        )}
        dataTestId={dataTestId}
        disabled={disabled}
        highlighted={highlighted}
        isInError={errorMessage !== undefined}
        options={inputValue}
        onChange={onChange}
        readOnly={readOnly}
      />
    </GenericField>
  );
};

CheckboxField.defaultProps = {
  disabled: false,
  errorMessage: undefined,
  fieldClassName: undefined,
  fieldSize: undefined,
  helperText: undefined,
  highlighted: false,
  inline: false,
  inputClassName: undefined,
  inputValue: undefined,
  isClearable: false,
  label: undefined,
  labelSize: undefined,
  mandatory: false,
  onChange: undefined,
  placeholder: undefined,
  readOnly: false,
};

export default CheckboxField;
