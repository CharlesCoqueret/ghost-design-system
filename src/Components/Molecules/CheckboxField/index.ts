export type { ICheckboxFieldProps } from './CheckboxField';
export { default as CheckboxField } from './CheckboxField';
