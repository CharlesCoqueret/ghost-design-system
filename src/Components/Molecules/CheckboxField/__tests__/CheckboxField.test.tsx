import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { CheckboxField } from '../CheckboxField';

describe('CheckboxField Component', () => {
  it('CheckboxField renders and handles changes', async () => {
    const onChangeMock = jest.fn();

    const { container } = render(
      <CheckboxField
        dataTestId={'DATA-TEST-ID'}
        input={[
          {
            label: 'Label 1',
            value: 'VALUE1',
          },
        ]}
        onChange={onChangeMock}
      />,
    );
    expect(container).toMatchSnapshot();

    const label = await screen.findByTestId('DATA-TEST-ID-0');
    userEvent.click(label);

    expect(container).toMatchSnapshot();
    expect(onChangeMock).toBeCalledTimes(1);
    expect(onChangeMock).toBeCalledWith([
      {
        checked: true,
        label: 'Label 1',
        value: 'VALUE1',
      },
    ]);
  });

  it('CheckboxField renders with values in readonly', () => {
    const { container } = render(
      <CheckboxField
        input={[
          {
            label: 'Label 1',
            value: 'VALUE1',
          },
        ]}
        readOnly
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('CheckboxField renders with values in disabled highligted', () => {
    const { container } = render(
      <CheckboxField
        input={[
          {
            label: 'Label 1',
            value: 'VALUE1',
            highlighted: true,
          },
          {
            label: 'Label 2',
            value: 'VALUE2',
            highlighted: false,
          },
          {
            label: 'Label 3',
            value: 'VALUE3',
          },
        ]}
        disabled
        highlighted
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('CheckboxField renders with values with fieldSize and inline', () => {
    const { container } = render(
      <CheckboxField
        input={[
          {
            label: 'Label 1',
            value: 'VALUE1',
            highlighted: true,
          },
        ]}
        inline
        errorMessage='ERROR'
        fieldSize={6}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
