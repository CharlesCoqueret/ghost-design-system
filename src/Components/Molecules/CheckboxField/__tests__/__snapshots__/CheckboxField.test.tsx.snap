// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxField Component renders and handles changes 1`] = `
<div>
  <div
    class="fieldGroup"
  >
    <div
      class="container"
    >
      <label
        class="label"
        data-testid="DATA-TEST-ID-0"
        for="checkbox-1"
        tabindex="0"
      >
        <div
          aria-label="Label 1"
          class="checkboxMarker"
          id="checkbox-1"
          role="checkbox"
        >
          <i
            class="icon"
            icon="fal,square"
          />
        </div>
        Label 1
      </label>
    </div>
  </div>
</div>
`;

exports[`CheckboxField Component renders and handles changes 2`] = `
<div>
  <div
    class="fieldGroup"
  >
    <div
      class="container"
    >
      <label
        class="label"
        data-testid="DATA-TEST-ID-0"
        for="checkbox-1"
        tabindex="0"
      >
        <div
          aria-label="Label 1"
          class="checkboxMarker"
          id="checkbox-1"
          role="checkbox"
        >
          <i
            class="icon"
            icon="fal,square"
          />
        </div>
        Label 1
      </label>
    </div>
  </div>
</div>
`;

exports[`CheckboxField Component renders with values in disabled highligted 1`] = `
<div>
  <div
    class="fieldGroup"
  >
    <div
      class="container"
    >
      <label
        class="label disabled highlighted"
        for="checkbox-3"
        tabindex="-1"
      >
        <div
          aria-label="Label 1"
          class="checkboxMarker disabled"
          id="checkbox-3"
          role="checkbox"
        >
          <i
            class="icon"
            icon="fas,square"
          />
        </div>
        Label 1
      </label>
      <label
        class="label disabled"
        for="checkbox-4"
        tabindex="-1"
      >
        <div
          aria-label="Label 2"
          class="checkboxMarker disabled"
          id="checkbox-4"
          role="checkbox"
        >
          <i
            class="icon"
            icon="fas,square"
          />
        </div>
        Label 2
      </label>
      <label
        class="label disabled"
        for="checkbox-5"
        tabindex="-1"
      >
        <div
          aria-label="Label 3"
          class="checkboxMarker disabled"
          id="checkbox-5"
          role="checkbox"
        >
          <i
            class="icon"
            icon="fas,square"
          />
        </div>
        Label 3
      </label>
    </div>
  </div>
</div>
`;

exports[`CheckboxField Component renders with values in readonly 1`] = `
<div>
  <div
    class="fieldGroup"
  >
    <div
      class="container"
    >
      <label
        class="label readOnly"
        for="checkbox-2"
        tabindex="-1"
      >
        <div
          aria-label="Label 1"
          class="checkboxMarker readOnly"
          id="checkbox-2"
          role="checkbox"
        >
          <i
            class="icon"
            icon="fas,square"
          />
        </div>
        Label 1
      </label>
    </div>
  </div>
</div>
`;

exports[`CheckboxField Component renders with values with fieldSize and inline 1`] = `
<div>
  <div
    class="fieldGroup size-6 inline"
  >
    <div
      class="container inlineContainer"
    >
      <label
        class="label inlineLabel"
        for="checkbox-6"
        tabindex="0"
      >
        <div
          aria-label="Label 1"
          class="checkboxMarker"
          id="checkbox-6"
          role="checkbox"
        >
          <i
            class="icon"
            icon="fal,square"
          />
        </div>
        Label 1
      </label>
    </div>
  </div>
</div>
`;
